### ISSUE TITLE
Setting hyperpriors on common effects

### ISSUE THREAD

**Issue by aswan89** [bambinos/bambi#889]
I'm attempting to run a synthetic control analysis where I regress a large number of pseudo-control subgroups against a pseudo-treatment group to assess the impact of the treatment. The basic structure of the model I'm trying to set up is something like `treatment ~ control1 + control2 + control3 + ... controln`. Each term, both the treatment and controls, represent the same type of quantity. The idea is to predict the value of `treatment` from the post-intervention subgroup behaviors and compare the prediction to the observed value.

Because I have a large number of predictors I wanted to do a penalized regression of some kind and came across [this paper](https://royalsocietypublishing.org/doi/10.1098/rsif.2018.0572#d1949703e1) which uses a Logit-Normal distribution to create a continuous analog of the Spike-and-Slab prior.

If I'm reading the paper right this can be expressed in Bambi like so:
```
bambi.Prior(
    'Normal',
    mu = 0,
    sigma = bambi.Prior(
        'LogitNormal',
        mu = 0, sigma = 10
    )
)
```

When I try to set this prior on my predictors I get the error `Trying to set hyperprior on 'control1'. Can't set a hyperprior on common effects.`

I'm somewhat new to bayesian modeling so I might be structuring this the wrong way, but it seems like Bambi is assuming that hyperpriors are only going to be used in hierarchical models. I'm pretty sure that I don't want a hierarchical model since I want the model coefficients to "blend" the predictor values such that they match the pre-treatment behavior of the treated subgroup. 

Am I structuring this all wrong or is this a limitation in how Bambi expects priors to be specified?

**Comment by tomicapretto** [comment]
@aswan89 thanks for the very well described issue. Do you have a snippet with a minimal version of the code you're using to define your model? 

What's the nature of the `control1` covariate? (numeric, categorical, indicator, etc.)

> Am I structuring this all wrong or is this a limitation in how Bambi expects priors to be specified?
Right now, we're limiting how users specify priors for common effects, but this may be an example indicating we need to make that more flexible.

**Comment by aswan89** [comment]
Here is the model construction:
```
import bambi

model_args = {
  "formula": "treatment ~ control1 + control2 + controln",
  "priors": {
    "common": bambi.Prior("Normal", mu = 0,
       sigma = bambi.Prior("LogitNormal", mu = 0, sigma = 10)
     ),
     "sigma": bambi.Prior("Exponential", lam = .5)
  },
  "family": "gaussian"
}
mod = bambi.Model(
    data = self.model_data,
    **model_args
)

mod_results = mod.fit()
```

Both the `treatment` and control variables are standardized numeric variables.

I was able to get it to run by going outside of bambi with a custom prior:

```
import pymc
import bambi

def ln_cass(name, mu = 0, sig = 10, tau = 5, *args, dims = None, **kwargs):
    lam = pymc.LogitNormal(name + '_mixture', mu = mu, sigma = sig)
    return pymc.Normal(name, mu = 0, sigma = tau*lam, *args, dims=dims, **kwargs)

model_args = {
  "formula": "treatment ~ control1 + control2 + controln",
  "priors": {
    "common": bambi.Prior('ln_cass', dist = ln_cass),
     "sigma": bambi.Prior("Exponential", lam = .5)
  },
  "family": "gaussian"
}
mod = bambi.Model(
    data = self.model_data,
    **model_args
)

mod_results = mod.fit()
```

This model did not have good convergence properties though, and I'm not certain if its because I malformed the ln_cass prior or misused it in the model. 

Based on some of the PyMC documentation I think I may need to incorporate elements from outside of the pymc Distributions namespace to properly express the ln-cass prior which would further complicate expressing it within bambi.

**Comment by tomicapretto** [comment]
The approach you're following, via custom priors, is what I would have done. If you're triying to reproduce the following prior:

![Image](https://github.com/user-attachments/assets/61dc42e5-b489-4ecd-b7a7-b97e647075c4)

then you're doing the correct thing. 
However:

* Do you think `sigma` equal to `10` is appropriate? Doesn't it put all probability mass near 0 or 1?
* Do you think `tau` equal to `5` is appropriate? That is quite a non-informative prior for a standardized variable. What you're saying is that the mean of the response can change between -15 and +15 standard deviations for a change equal to a standard deviation in the predictors.

Two things you can do to double-check you're doing things correctly

* Reproduce the simulation study in the paper (or some subset of it)
* Write the model both in Bambi and PyMC and compare results (I can imagine that you're using Bambi because you want to avoid writing a lot of PyMC code, so this may not be the preferred approach)

**Comment by aswan89** [comment]
I took the values for tau and sigma from this paragraph in the discussion section:

> Our prior requires the choices of three hyperparameters, although we contend that they are much more interpretable than those required for other Bayesian shrinkage methods (see the electronic supplementary material, sections S1 and S2 for further detail on the roles of the hyperparameters). The three hyperparameters required correspond to, firstly, the standard deviation of the ‘slab’ component, which we refer to as τ; for standardized predictors, a default value of τ = 5 has been sufficient for all of our applications because it essentially provides a vague Gaussian prior for non-zero coefficients. Secondly, the parameters of the logit-normal distribution ([figure 1](https://royalsocietypublishing.org/doi/10.1098/rsif.2018.0572#RSIF20180572F1)a) must be specified; we refer to these parameters as 𝜇𝜆 and 𝜎𝜆. 𝜇𝜆 can be chosen based on our prior beliefs about the probability of a zero coefficient, and in our experience does not require much tuning; the median of the logit-normal distribution is given by sigm⁡(𝜇𝜆), where sigm⁡(⋅) is the logistic sigmoid function. Thus, if we believe a priori that each coefficient has a probability a of being non-zero, we simply set 𝜇𝜆=logit⁡(𝑎). 𝜎𝜆
 simply controls the quality of the approximation to the spike-and-slab prior, with larger values corresponding to better approximations. We have used a default value of 𝜎𝜆=10 throughout the paper; results are not sensitive to increases in this value.

Reproducing the simulation seems like a good next step. As to whether bambi needs to change to accommodate these types of complex priors, it seems like the custom prior solution makes more sense than stacking instances of `bambi.Prior` together.

**Comment by tomicapretto** [comment]
We could eventually incorporate this prior into Bambi itself, doing something similar to what we did for the horseshoe. 
https://github.com/bambinos/bambi/blob/a798c89be873987eadf30dba021b5f030cecf011/bambi/backend/utils.py#L9-L37

**Comment by ivlis** [comment]
Hi @aswan89 , i'm trying to use bambi for synthetic control as well. In the classical synthetic control your coefficients should sum up to 1.  Have you tried to use Dirichlet prior? This will automatically disallow the model to overfit to only one "control" subject.
